src/tracker/embedder/embedder.py

```diff
@@ -0,0 +1,102 @@
+from abc import ABC, abstractmethod
+from typing import List
+
+import numpy as np
+import torch
+
+from src.config.config import EmbedderConfig
+from src.image.image import ImageObject
+from src.tracker.detector.detection import Detection
+from src.tracker.utils import (
+    pad_image_to_target_size,
+    resize_for_padding,
+)
+
+
+class Embedder(ABC):
+    def __init__(self, cfg: EmbedderConfig):
+        """
+        Initializes the detector with a DetectorConfig object that contains configuration for the model.
+        """
+        self.cfg = cfg
+        self.distance = cfg.embedder_distance
+        self.input_height = cfg.input_height
+        self.input_width = cfg.input_width
+        self.input_shape = (cfg.input_height, cfg.input_width)
+
+    @abstractmethod
+    async def compute_features(
+        self, image: ImageObject, detections: List[Detection]
+    ) -> List[np.ndarray]:
+        """
+        Abstract method to be implemented by specific detector classes. Each detector must implement
+        this method to detect objects in the provided image.
+        """
+        pass
+
+    def compute_distance(
+        self, feature_vector_1: torch.Tensor, feature_vector_2: torch.Tensor
+    ):
+        """
+        Compute pairwise distances between feature vectors using PyTorch.
+
+        :param feature_vector_1: First feature vector (PyTorch tensor).
+        :param feature_vector_2: Second feature vector (PyTorch tensor).
+        :return: Computed distance.
+        """
+        if self.distance == "euclidean":
+            distance = torch.norm(feature_vector_1 - feature_vector_2, p=2)
+        elif self.distance == "cosine":
+            distance = 1 - torch.nn.functional.cosine_similarity(
+                feature_vector_1.unsqueeze(0), feature_vector_2.unsqueeze(0)
+            )
+        elif self.distance == "manhattan":
+            distance = torch.sum(torch.abs(feature_vector_1 - feature_vector_2))
+        else:
+            raise ValueError(f"Unsupported metric '{self.distance}'")
+        return distance.cpu().item()
+
+    def crop_detections(
+        self, image: ImageObject, detections: List[Detection]
+    ) -> torch.Tensor:
```

Comment:
I think this should return a np.array and not a torch.Tensor.  Still let it be a stack of image tensors but we should keep it general

Reply:
done, maybe the default embedder will expect another input typ though.. For now, I changed to np.array as you said and when we have the default embedder we can make this class abstract and implement it differently depending which embedder we use